File: backend/app/models/analysis.py
"""
Analysis models for AI transparency and confidence scoring.

These models support EPIC Q: AI Transparency & UX Polish by providing
comprehensive transparency information about AI analysis decisions.
"""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class SimilarIssueRef(BaseModel):
    """Reference to a similar issue used in analysis."""

    id: int = Field(..., description="Issue ID in Sentry")
    title: str = Field(..., description="Issue title")
    similarity: float = Field(..., ge=0.0, le=1.0, description="Similarity score (0-1)")
    project: Optional[str] = Field(None, description="Project name")
    status: Optional[str] = Field(None, description="Issue status")
    first_seen: Optional[datetime] = Field(None, description="First occurrence")
    last_seen: Optional[datetime] = Field(None, description="Last occurrence")


class ModelInfo(BaseModel):
    """Information about the AI model used."""

    name: str = Field(..., description="Model name")
    provider: str = Field(..., description="Provider (ollama, openai, anthropic)")
    version: Optional[str] = Field(None, description="Model version")
    context_window: Optional[int] = Field(None, description="Context window size")


class ConfidenceFactors(BaseModel):
    """Factors contributing to confidence score."""

    high_similarity_count: int = Field(
        ..., description="Number of similar issues with >0.8 similarity"
    )
    enrichment_coverage: float = Field(
        ..., ge=0.0, le=1.0, description="Percentage of enrichment sources available (0-1)"
    )
    freshness_score: float = Field(
        ..., ge=0.0, le=1.0, description="Weighted freshness of data used (0-1)"
    )
    reasoning: str = Field(..., description="Human-readable confidence reasoning")


class TransparencyInfo(BaseModel):
    """Transparency information about AI analysis."""

    similar_issues_count: int = Field(..., description="Total similar issues analyzed")
    similar_issues: List[SimilarIssueRef] = Field(
        default_factory=list, description="Similar issues used in analysis"
    )
    enrichment_sources_used: List[str] = Field(
        default_factory=list, description="Enrichment data sources included"
    )
    enrichment_sources_stale: List[str] = Field(
        default_factory=list, description="Enrichment sources with stale data"
    )
    ranking_variant: str = Field(
        default="multi-signal", description="Ranking algorithm variant used"
    )
    model_info: ModelInfo = Field(..., description="AI model information")
    confidence_factors: ConfidenceFactors = Field(..., description="Confidence breakdown")


class AnalysisSources(BaseModel):
    """Sources used in analysis."""

    similar_issues: List[SimilarIssueRef] = Field(default_factory=list)
    enrichment_data: dict = Field(default_factory=dict)
    citations: List[str] = Field(default_factory=list)


class PIIScrubResult(BaseModel):
    """Result of PII scrubbing operation with tracking."""

    scrubbed_data: dict = Field(..., description="Data after PII scrubbing")
    pii_detected: bool = Field(..., description="Whether PII was found and scrubbed")
    fields_scrubbed: List[str] = Field(
        default_factory=list, description="Field paths that were scrubbed"
    )
    scrub_count: int = Field(default=0, description="Total number of scrub operations")


class AnalysisResponse(BaseModel):
    """Complete AI analysis response with transparency."""

    analysis: str = Field(..., description="AI-generated analysis text")
    confidence: float = Field(..., ge=0.0, le=1.0, description="Confidence score (0-1)")
    sources: AnalysisSources = Field(..., description="Sources used in analysis")
    transparency: TransparencyInfo = Field(..., description="Transparency information")
    processing_time_ms: float = Field(..., description="Processing time in milliseconds")
    pii_scrub_info: Optional[PIIScrubResult] = Field(
        None, description="PII scrubbing information if applicable"
    )


class QuickPromptType(BaseModel):
    """Quick prompt template definition."""

    prompt_type: str = Field(..., description="Prompt type identifier")
    label: str = Field(..., description="Display label")
    icon: str = Field(..., description="Icon name")
    system_prompt: str = Field(..., description="System prompt template")
    enrichment_priority: List[str] = Field(
        default_factory=list, description="Enrichment sources to prioritize"
    )


class QuickPromptRequest(BaseModel):
    """Request for quick prompt analysis."""

    issue_id: int = Field(..., description="Issue ID to analyze")
    prompt_type: str = Field(..., description="Quick prompt type to use")
    user_query: Optional[str] = Field(None, description="Optional user query override")


class DataQualityIndicators(BaseModel):
    """Data quality indicators for UI display."""

    stale_sources: List[str] = Field(default_factory=list, description="Stale data sources")
    very_stale_sources: List[str] = Field(
        default_factory=list, description="Very stale data sources (critical)"
    )
    missing_sources: List[str] = Field(
        default_factory=list, description="Expected but missing sources"
    )
    pii_scrubbed: Optional[PIIScrubResult] = Field(None, description="PII scrubbing info")
    overall_quality_score: float = Field(
        ..., ge=0.0, le=1.0, description="Overall data quality (0-1)"
    )

File: frontend/src/types/analysis.ts
/**
 * Type definitions for AI analysis transparency and confidence scoring.
 *
 * These types support EPIC Q: AI Transparency & UX Polish by providing
 * comprehensive type safety for transparency information displayed to users.
 */

export interface SimilarIssueRef {
  id: number;
  title: string;
  similarity: number; // 0-1
  project?: string;
  status?: string;
  first_seen?: string;
  last_seen?: string;
}

export interface ModelInfo {
  name: string;
  provider: string;
  version?: string;
  context_window?: number;
}

export interface ConfidenceFactors {
  high_similarity_count: number;
  enrichment_coverage: number; // 0-1
  freshness_score: number; // 0-1
  reasoning: string;
}

export interface TransparencyInfo {
  similar_issues_count: number;
  similar_issues: SimilarIssueRef[];
  enrichment_sources_used: string[];
  enrichment_sources_stale: string[];
  ranking_variant: string;
  model_info: ModelInfo;
  confidence_factors: ConfidenceFactors;
}

export interface AnalysisSources {
  similar_issues: SimilarIssueRef[];
  enrichment_data: Record<string, any>;
  citations: string[];
}

export interface PIIScrubResult {
  scrubbed_data: Record<string, any>;
  pii_detected: boolean;
  fields_scrubbed: string[];
  scrub_count: number;
}

export interface AnalysisResponse {
  analysis: string;
  confidence: number; // 0-1
  sources: AnalysisSources;
  transparency: TransparencyInfo;
  processing_time_ms: number;
  pii_scrub_info?: PIIScrubResult;
}

export interface QuickPromptType {
  prompt_type: string;
  label: string;
  icon: string;
  system_prompt: string;
  enrichment_priority: string[];
}

export interface QuickPromptRequest {
  issue_id: number;
  prompt_type: string;
  user_query?: string;
}

export interface DataQualityIndicators {
  stale_sources: string[];
  very_stale_sources: string[];
  missing_sources: string[];
  pii_scrubbed?: PIIScrubResult;
  overall_quality_score: number; // 0-1
}

/**
 * Helper function to format confidence score as a level
 */
export function getConfidenceLevel(confidence: number): 'high' | 'medium' | 'low' {
  if (confidence >= 0.7) return 'high';
  if (confidence >= 0.4) return 'medium';
  return 'low';
}

/**
 * Helper function to get confidence color for Mantine components
 */
export function getConfidenceColor(confidence: number): string {
  const level = getConfidenceLevel(confidence);
  switch (level) {
    case 'high':
      return 'green';
    case 'medium':
      return 'yellow';
    case 'low':
      return 'red';
  }
}

/**
 * Helper function to format enrichment source names for display
 */
export function formatSourceName(source: string): string {
  return source
    .split('_')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
}

/**
 * Helper to check if data is stale (older than 7 days)
 */
export function isStale(fetchedAt: string | undefined): boolean {
  if (!fetchedAt) return true;

  const fetched = new Date(fetchedAt);
  const now = new Date();
  const daysDiff = (now.getTime() - fetched.getTime()) / (1000 * 60 * 60 * 24);

  return daysDiff > 7;
}

/**
 * Helper to check if data is very stale (older than 30 days)
 */
export function isVeryStale(fetchedAt: string | undefined): boolean {
  if (!fetchedAt) return true;

  const fetched = new Date(fetchedAt);
  const now = new Date();
  const daysDiff = (now.getTime() - fetched.getTime()) / (1000 * 60 * 60 * 24);

  return daysDiff > 30;
}
